### PR TITLE
bug(materials validation)

### LINKED ISSUE #45: Materials error tag doesn't make sense
## Problem
When some resources of materials was deleted. It also show the resources and show error tag `no materials`. So, admin have to reset resource before (choose '-' in selected box ) and add new resource. If admin forget reset them. It also keep deleted resources in materials.

img_21022018_105332_0
  
So, I will propose solution follow this.

1. Show (x) icon in material column of rejected products table.
2. Add/ Repairs rejected product.
3. Remove deleted materials from product.
4. Validate product and show `* removed deleted materials`  in case which there are some materials was deleted.
5. User save product.
6. Product isn't in the table of rejected products.

### PR DESCRIPTION
**Pull request by lalitaCF**
close #45

### CODE REVIEW

**Review by Solomonic — CHANGES_REQUESTED** [review]
Changes as discussed

File: assets/app.js [hunk 1]
```diff
@@ -338,6 +338,39 @@ new Vue({
         }
     },
     computed: {
+        validateCacheMaterial: function () {
```
Comment (Solomonic) [inline assets/app.js]: This is not how you use **computed** properties, you named, coded and use it like a normal **method**
please move.

File: assets/app.js [hunk 2]
```diff
@@ -338,6 +338,39 @@ new Vue({
...
+            active_languages = this.activeLanguagesId
```
Comment (Solomonic) [inline assets/app.js]: when you use it as a normal method, do it for ALL languages

File: assets/app.js [hunk 3]
```diff
@@ -338,6 +338,39 @@ new Vue({
...
+            this.products.forEach(function (product) {
```
Comment (Solomonic) [inline assets/app.js]: can we do it when we add/get product(s)?

File: assets/app.js [hunk 4]
```diff
@@ -338,6 +338,39 @@ new Vue({
...
+                    });
+
+                    is_overridden = true
+                    if (materials.length === 0) {
+                        is_overridden = false
+                    }
+                    //compare cached_materials and materials
+                    if (cached_materials.length === materials.length) {
+                        is_overridden = false
+                        for (i = 0; i < materials.length; i++) {
```
Comment (Solomonic) [inline assets/app.js]: is there a reason for `for (i = 0; i < materials.length; i++)` instead `materials.foreach(...)`?

File: assets/app.js [hunk 5]
```diff
@@ -794,39 +825,54 @@ new Vue({
...
+            let verified_resources = {
+                'base_product': true,
+                'component1': true,
+                'component2': true,
+                'materials': true,
+                'no_deleted_materials': true
+            }
+            for (var type in verified_resources) {
+                if (type === 'no_deleted_materials') {
+                    continue
```
Comment (Solomonic) [inline assets/app.js]: ???

Comment (Solomonic) [reply assets/app.js]: If you don't want to loop it, don't put it into `verified_resources` in the beginning, just add it in the end when you know if its true/false

File: assets/app.js [hunk 6]
```diff
@@ -794,39 +825,54 @@ new Vue({
...
+                    if (product[type].length === 0) {
+                        verified_resources.materials = false
+                    }
+                    if (product.deleted_materials_resources.length > 0 && product.initial_materials_amount > 0) {
+                        product.dirty = true
+                        verified_resources['no_deleted_materials'] = false
+                    }
+                }
+            }
+            return verified_resources
```
Comment (Solomonic) [inline assets/app.js]: all the part above needs refactoring!
just a couple of things

-  don't repeat yourself! you check 5 times if one product is rejected
- if you know that `type ==materials/component/base_product` in a if condition please use it: `product['materials']` is then better then `product[type]`: easier to read and slightly faster

**Review by Solomonic — CHANGES_REQUESTED** [review]

File: assets/app.js (hunk 3 above)
Comment (lalita13008) [reply assets/app.js]: `foreach` cannot use `break`

File: assets/app.js (hunk 4 above)
Comment (lalitaCF) [reply assets/app.js]: `foreach` cannot use `break`

File: assets/app.js (hunk 3 above)
Comment (lalitaCF) [reply assets/app.js]: it active when we have data in products and open materials page.

File: assets/app.js (hunk 2 above)
Comment (lalitaCF) [reply assets/app.js]: Why?

File: assets/app.js (hunk 4 above)
Comment (cyberhck) [reply assets/app.js]: maybe you'd like `materials.some` not sure what's it doing, so.

interesting fact about `foreach`, `map`, `filter`, etc.
In some languages, not js obviously, they don't guarantee the order of what it's gonna be called in.
for example, if you're doing map, across 1000 items, your compiler can break it down into 4 parts, then do a map on that in a parallel way (here 4 being number of core in processor, for example), and collect it back together later. they'll finally have the same order, but doing side effects inside that functions are frowned upon.

:smile: I watched a functional programming talk recently.

File: assets/app.js [hunk 7]
```diff
@@ -794,39 +788,88 @@ new Vue({
         },
     },
     methods: {
-        validateRejectedProduct: function (type, product) {
-            if (type == 'materials' && product['is_rejected'] && product['rejected_' + type]) {
-                if (!(product['is_rejected'] && product['rejected_' + type])) {
-                    return true
+        validateCacheMaterial: function () {
+            active_languages = this.activeLanguagesIds
+            this.products.forEach(function (product) {
```
Comment (Solomonic) [inline assets/app.js]: What about something like:
```
 this.products.forEach(function (product) {
    active_languages.forEach(function (language) {
       // we assume that deleted materials are already removed from product.materials
        current_materials[language] = product.materials[language].join(" / ")
        if(current_materials[language]===product.cached_materials[language]){
             product.cached_materials[language]['is_overridden'] = false
        }
        else{
            product.cached_materials[language]['is_overridden'] = false
        }
    }
}
```

it should do pretty much the same thing or not?

File: assets/app.js [hunk 8]
```diff
@@ -794,39 +788,88 @@ new Vue({
...
+        },
+        validateRejectedProduct: function (product) {
+            let verified_resources = {
+                'base_product': true,
+                'component1': true,
+                'component2': true,
+                'materials': true
+            }, no_deleted_materials = true
+
+            if (!product['is_rejected']) {
```
Comment (Solomonic) [inline assets/app.js]: i don't think we need this check. I think this method is more useful when it also validate not-rejected products

File: assets/app.js [hunk 9]
```diff
@@ -794,39 +788,88 @@ new Vue({
...
+                verified_resources['no_deleted_materials'] = no_deleted_materials
+                return verified_resources
+            }
+
+            for (var type in verified_resources) {
+                if (type === 'base_product' && product['rejected_base_product']) {
```
Comment (Solomonic) [inline assets/app.js]: I don't like that we rely on _['rejected_base_product']_ 
checking should be easy enough:
```
if (type === 'base_product' && (!product['base_product'] || !product['base_product']['active'] ){
    verified_resources['base_product'] = false
}

File: assets/app.js [hunk 10]
```diff
@@ -1276,7 +1319,7 @@ new Vue({
                     });
             }
         },
-        prepareProducts: function(raw_products, is_rejected_products = false) {
+        prepareProducts: function(raw_products) {
```
Comment (Solomonic) [inline assets/app.js]: remove deleted base_products, components, materials, metatags from each product

**Review by Solomonic — CHANGES_REQUESTED** [review]

File: assets/app.js (hunk 7 above)
Comment (cyberhck) [reply assets/app.js]: the inner loop, why not: `product.cached_materials[language]['is_overridden'] = !current_materials[language]===product.cached_materials[language]` :slightly_smiling_face:

File: assets/app.js [hunk 11]
```diff
@@ -758,6 +754,7 @@ new Vue({
             }
 
             this.showLoading = true
+            this.optionsMaterial
```
Comment (Solomonic) [inline assets/app.js]: ????
remove or explain

File: assets/app.js [hunk 12]
```diff
@@ -791,44 +786,112 @@ new Vue({
...
+        }
     },
     methods: {
-        validateRejectedProduct: function (type, product) {
-            if (type == 'materials' && product['is_rejected'] && product['rejected_' + type]) {
-                if (!(product['is_rejected'] && product['rejected_' + type])) {
-                    return true
-                }
+        getResourcesName:function (resources) {
+            if (resources !== null) {
```
Comment (Solomonic) [inline assets/app.js]: better check if resources is an array:
```
if(Array.isArray(resources)){ ... }
else return []
```

File: assets/app.js (hunk 4 above)
Comment (Solomonic) [reply assets/app.js]: this is a very complicated way to figure out if something changed. I don't think this works properly.
What about
```
// filter active materials
current_materials_list = product.materials.map(function (material) {
   if(material.is_active) return material[language]
}
// join materials to preview what we'll write back to DB
current_materials = current_materials_list.join("/") 
product.cached_materials[language]['is_overridden'] = (product.cached_materials[language]=== current_materials)

**Review by Solomonic — CHANGES_REQUESTED** [review]

File: assets/app.js [hunk 13]
```diff
@@ -791,59 +785,128 @@ new Vue({
...
+        isActiveBaseProduct: function (name) {
+            return this.getResourcesName(this.translationsBaseProducts).indexOf(name) > -1
+        },
+        isActiveComponents: function (name) {
+            return this.getResourcesName(this.translationsComponents).indexOf(name) > -1
+        },
+        isActiveMaterials: function (name) {
+            return this.getResourcesName(this.translationsMaterials).indexOf(name) > -1
+        },
+        showErrorLabelBaseProduct: function (product) {
```
Comment (Solomonic) [inline assets/app.js:831]: shorten to one-liner with ternary operator 
```
return ( this.isEmptyResource(product['base_product']) || 
             !this.isActiveBaseProduct(product['base_product']['name'])) ? true : false
```

File: assets/app.js [hunk 14]
```diff
@@ -791,59 +785,128 @@ new Vue({
...
         },
-        isRejectedProduct: function (product) {
-            if (product['base_product'] == null || Object.keys(product['base_product']).length === 0 || product['base_product'].length === 0) {
+        showErrorLabelComponent1: function (product) {
+            if (this.isEmptyResource(product['component1'])) {
+                // there is component2 but no component1
+                return !(this.isEmptyResource(product['component2']) || product['component2'].name === '-')
+            }
+            // verify component1 was deleted after prepareProduct().
+            if (product['component1'].deleted) {
```
Comment (Solomonic) [inline assets/app.js]:
```
return ( product['component1'].deleted || !this.isActiveComponents(product['component1']['name'] )? true: false

File: assets/app.js [hunk 15]
```diff
@@ -791,59 +785,128 @@ new Vue({
...
+            // verify component1 when it was deleted from translations page
+            if (!this.isActiveComponents(product['component1']['name'])) {
                 return true
             }
 
-            if (product['component1'] == null || Object.keys(product['component1']).length === 0 || product.length === 0) {
-                return !(product['component2'] == null || Object.keys(product['component2']).length === 0 || product.length === 0 || product['component2'].name === '-')
+            return false
+        },
+        showErrorLabelComponent2: function (product) {
```
Comment (Solomonic) [inline assets/app.js]:
```
component = product['component2'];
return 
(!(this.isEmptyResource(component)) && component.deleted) || !this.isActiveComponents(component['name']))
? true :false
```

File: assets/app.js [hunk 16]
```diff
@@ -791,59 +785,128 @@ new Vue({
...
+            if (!(this.isEmptyResource(product['component2'])) && product['component2'].deleted) {
+                return true
+            }
+            // verify component2 when it was deleted from translations page
+            if (!this.isActiveComponents(product['component2']['name'])) {
+                return true
+            }
+            return false
+        },
+        showErrorLabelMaterials: function (product) {
```
Comment (Solomonic) [inline assets/app.js]:
```
return (this.isEmptyResource(product['materials']))? true : false
```

File: assets/app.js [hunk 17]
```diff
@@ -791,59 +785,128 @@ new Vue({
...
+                    break
+                }
             }
 
             return false
         },
-        fetchResource: function(type, variable_name){
+        isRejectedProduct: function (product) {
+            product.has_deleted_materials = false
+            if (this.showErrorLabelBaseProduct(product) ||
```
Comment (Solomonic) [inline assets/app.js]:
```
return (
  this.showErrorLabelBaseProduct(product) ||
  this.showErrorLabelComponent1(product) ||
  this.showErrorLabelComponent2(product) ||
  this.showErrorLabelMaterials(product)
) ? true : false
```

**Review by Solomonic — CHANGES_REQUESTED** [review]

**Review by Solomonic — APPROVED** [review]

**Pull request merged**